### PR TITLE
Mark and sweep strategy for archival and deletion

### PR DESCRIPTION
**Pull request by Ladas**
Allowing mark&sweep for archival and deletion

First we send parts, client needs to generate :refresh_state_uuid, :refresh_state_part_uuid. Where :refresh_state_uuid must be same for all the parts and :refresh_state_part_uuid. We do expect uuid format.

After all parts were sent we send :total_parts, which is integer marking how many parts we expect with refresh_state_uuid to identify to which full refresh this belongs to. And optionally :sweep_scope that allows us to do full refresh of just 1 or more inventory collection.

The sweeping waits until all parts were saved, then it deletes/archives all records that were not touched (we use :last_seen_on column for this)

### CODE REVIEW

**Comment by agrare** [comment]
@Ladas can you document how this works in the description?

**Comment by Ladas** [comment]
@agrare yes,  I'll comment it and clean it up next week.

File: lib/inventory_refresh/inventory_collection.rb [hunk 1]
```diff
@@ -206,7 +209,19 @@ def store_updated_records(records)
     #
     # @param records [Array<ApplicationRecord, Hash>] list of stored records
     def store_deleted_records(records)
-      @deleted_records.concat(records_identities(records))
+      # TODO(lsmola) we need to move output streaming here, output_client of manageiq-messaging should be passed here
+      # and we should just keep track of a count here, not all ids. At least for deleting, the ids do not have upper
+      # limit, so the memory could bump a lot.
```
Comment (agrare) [inline lib/inventory_refresh/inventory_collection.rb]: We could yield the created/updated/deleted records in batches to the caller, just cutting off the deleted records at some fixed max doesn't seem right

Comment (agrare) [reply lib/inventory_refresh/inventory_collection.rb]: Or passing in a manageiq-messaging client, since then we're baking in how we're calling the object passed in in a way that won't be very applicable to other types of output clients

File: lib/inventory_refresh/inventory_collection.rb [hunk 2]
```diff
@@ -564,7 +579,9 @@ def db_collection_for_comparison_for(references)
     # @return [ActiveRecord::Relation] relation that can fetch all the references from the DB
     def full_collection_for_comparison
       return arel unless arel.nil?
-      parent.send(association)
+      rel = parent.send(association)
+      rel = rel.active if rel && supports_column?(:archived_on) && retention_strategy == :archive
```
Comment (agrare) [inline lib/inventory_refresh/inventory_collection.rb]: `supports_column?(:archived_on)` Looks like another place where we should have the archived_timestamp_column on the IC

Comment (agrare) [reply lib/inventory_refresh/inventory_collection.rb]: I also wonder if we shouldn't be calling `.active` on a relation since that assumes a lot as a gem.  I'd prefer the association we pass in only include the actives (and have e.g. all_vms/all_orch_stacks which includes the inactives) in the model

File: lib/inventory_refresh/logging.rb [hunk 3]
```diff
@@ -4,7 +4,13 @@ class << self
...
 
   def self.logger
-    @logger ||= NullLogger.new
+    @logger ||= default_logger
+  end
+
+  def self.default_logger
+    logger = Logger.new(STDOUT)
+    logger.level = Logger::ERROR
+    logger
```
Comment (agrare) [inline lib/inventory_refresh/logging.rb]: Is this related to mark&sweep?  If you want to pass in a logger for tests lets do that not change the default logger.

File: lib/inventory_refresh/inventory_collection.rb (hunk 1 above)
Comment (Ladas) [reply lib/inventory_refresh/inventory_collection.rb]: Yeah I have the TODO there, we'll need to setup the streaming topic and push it there in batches right here, so the memory doesn't grow. I think it should be ok to have this as temporary solution?

That means passing the Streams objects to persister and IC.

Comment (agrare) [reply lib/inventory_refresh/inventory_collection.rb]: Lets drop it from here since it isn't related to mark&sweep

File: lib/inventory_refresh/inventory_collection.rb (hunk 2 above)
Comment (Ladas) [reply lib/inventory_refresh/inventory_collection.rb]: right, I am not sure what is better. If we'd do it explicitly, we'd need to define each relation twice in the Source (and for through methods that would spread to every model?)

File: lib/inventory_refresh/logging.rb (hunk 3 above)
Comment (Ladas) [reply lib/inventory_refresh/logging.rb]: well I'd like to see at least ERRs as default, that is for specs and also for dev, otherwise it's confusing not to see errors anywhere :-)

not sure if we can make it default per ENV, still don't have anything like Rails.env, right?

Comment (agrare) [reply lib/inventory_refresh/logging.rb]: I don't expect a gem I include to start writing to STDOUT

Comment (Ladas) [reply lib/inventory_refresh/logging.rb]: so what would be the place to put it on, so it works as default for the gems specs and for dev?

File: lib/inventory_refresh/inventory_collection.rb (hunk 1 above)
Comment (Ladas) [reply lib/inventory_refresh/inventory_collection.rb]: ok

File: lib/inventory_refresh/logging.rb (hunk 3 above)
Comment (agrare) [reply lib/inventory_refresh/logging.rb]: You could set the logger in the spec_helper and bin/console for dev.  For any apps including the gem we already set the logger or should if we don't

Comment (Ladas) [reply lib/inventory_refresh/logging.rb]: ok, will do

File: lib/inventory_refresh/persister/mark_and_sweep_mixin.rb [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      extend ActiveSupport::Concern
+
+      included do
+        include InventoryRefresh::SaveCollection::Saver::SqlHelper
+      end
+
+      private
+
+      def define_refresh_state_ics
+        refresh_states_inventory_collection = ::InventoryRefresh::InventoryCollection.new(
```
Comment (agrare) [inline lib/inventory_refresh/persister/mark_and_sweep_mixin.rb]: This looks like it belongs in the app not in the gem, we're forcing any app that includes this to define models with specific names to handle this.

The calling app should handle the refresh state/parts and just pass in the delete conditions to the gem IMO

Comment (Ladas) [reply lib/inventory_refresh/persister/mark_and_sweep_mixin.rb]: hm, we'd probably need something in the middle

Comment (Ladas) [reply lib/inventory_refresh/persister/mark_and_sweep_mixin.rb]: like a Persister gem :-)

File: lib/inventory_refresh/inventory_collection.rb (hunk 2 above)
Comment (agrare) [reply lib/inventory_refresh/inventory_collection.rb]: Should we only do this if there's a `last_seen_on` column?

File: lib/inventory_refresh/persister/mark_and_sweep_mixin.rb (hunk 4 above)
Comment (agrare) [reply lib/inventory_refresh/persister/mark_and_sweep_mixin.rb]: Haha like some sort of ... topological_inventory-persister?

File: lib/inventory_refresh/inventory_collection.rb (hunk 2 above)
Comment (Ladas) [reply lib/inventory_refresh/inventory_collection.rb]: Do do you want me to remove this, or lets keeps this until we do the scopes? (right now, this might spike memory for subcollections. For now we can observe the spike with mock collector)
